packages/compiler/core/checker.ts

```diff
@@ -1960,6 +2220,703 @@ export function createChecker(program: Program): Checker {
+    }
+    const kind =
+      typeof value === "string" ? "String" : typeof value === "number" ? "Number" : "Boolean";
+    const type: StringLiteralType | NumericLiteralType | BooleanLiteralType = createType({
+      kind,
+      value: value as any,
+      node: node as any, // todo: synthesized strings don't have a node...
+    });
+
+    program.literalTypes.set(value, type);
+    return type;
+  }
+
+  function evalProjectionDecoratorReference(
+    node: ProjectionDecoratorReferenceExpressionNode
+  ): Type {
+    const ref = resolveTypeReference(node.target, true);
+    if (!ref) throw new ProjectionError("Can't find decorator.");
+    compilerAssert(ref.kind === "decorator", "should only resolve decorator symbols");
+    return createType({
+      kind: "Function",
+      call(...args: Type[]): Type {
+        let retval = ref.value(program, ...marshalProjectionArguments(args));
+        return voidType;
+      },
+    } as const);
+  }
+
+  function evalProjectionIdentifier(node: IdentifierNode): Type {
+    // first check the eval context
+
+    let currentContext = evalContext;
+    while (currentContext) {
+      if (currentContext.locals.has(node.sv)) {
+        return currentContext.locals.get(node.sv)!;
+      }
+      currentContext = currentContext.parent;
+    }
+
+    // next, resolve outside
+    const ref = resolveTypeReference(node);
+    if (!ref) throw new ProjectionError("Unknown identifier " + node.sv);
+
+    switch (ref.kind) {
+      case "decorator":
+        // shouldn't ever resolve a decorator symbol here (without passing
+        // true to resolveTypeReference)
+        return errorType;
+      case "function":
+        // TODO: store this in a symbol link probably?
+        const t: FunctionType = createType({
+          kind: "Function",
+          call(...args: Type[]): Type {
+            let retval = ref.value(program, ...marshalProjectionArguments(args));
+            return marshalProjectionReturn(retval);
+          },
+        } as const);
+        return t;
+      case "type":
+      case "projection":
+        const links = getSymbolLinks(ref);
+        compilerAssert(links.declaredType, "Should have checked all types by now");
+
+        return links.declaredType;
+    }
+  }
+
+  function marshalProjectionArguments(args: Type[]): (Type | number | boolean | string | object)[] {
+    return args.map((arg) => {
+      if (arg.kind === "Boolean" || arg.kind === "String" || arg.kind === "Number") {
+        return literalTypeToValue(arg);
+      }
+      return arg;
+    });
+  }
+
+  function marshalProjectionReturn(value: unknown): Type {
+    if (typeof value === "boolean" || typeof value === "string" || typeof value === "number") {
+      return createLiteralType(value);
+    }
+
+    if (typeof value === "object" && value !== null) {
+      if ("kind" in value) {
+        return value as Type;
+      } else {
+        // this could probably be more robust
+        return createType({
```

Comment:
todoish?

Reply:
I think it's fine for now but if someone is digging around in here wondering why unexpected things are happening the comment could be helpful? Can also delete the comment.